cli/rack/__init__.py

```diff
@@ -350,12 +351,162 @@ def utility_copygraph_driver(base_url: Url, triple_store: Optional[Url], triple_
+                step['csv'] = path.name
+
+        self.data_graphs.add(obj['data-graph'])
+
+        c = obj.get('extra-data-graphs')
+        if c is not None:
+            self.data_graphs.update(c)
+
+        c = obj.get('model-graphs')
+        if c is not None:
+            if type(c) == str:
+                self.model_graphs.add(c)
+            elif type(c) == list:
+                self.model_graphs.update(c)
+
+        with open(to_path, mode='w', encoding='utf-8-sig', newline='\n') as out: 
+            yaml.safe_dump(obj, out)
+
+    def nodegroups(
+        self,
+        from_path: Path,
+        to_path: Path,
+    ) -> None:
+        shutil.copyfile(from_path.joinpath('store_data.csv'), to_path.joinpath('store_data.csv'), follow_symlinks=True)
+        with open(from_path.joinpath('store_data.csv'), 'r') as f:
+            for row in csv.DictReader(f):
+                json = row['jsonFile']
+                shutil.copyfile(from_path.joinpath(json), to_path.joinpath(json), follow_symlinks=True)
+
+    def manifest(
+        self,
+        from_path: Path,
+        to_path: Path,
+    ) -> None:
+
+        with open(from_path, mode='r', encoding='utf-8-sig') as f:
+            obj = yaml.safe_load(f)
+
+        base_path = from_path.parent
+        for i in range(len(obj['steps'])):
+            step = obj['steps'][i]
+            if 'manifest' in step:
+                path = step['manifest']
+                dirname = Path(f'{self.next_fresh():02}_manifest')
+                subdir = to_path.parent.joinpath(dirname)
+                topath = subdir.joinpath(Path(path).name)
+                mkdir(subdir)
+                self.manifest(base_path.joinpath(path), topath)
+                step['manifest'] = str(dirname.joinpath(Path(path).name))
+            elif 'model' in step:
+                path = step['model']
+                dirname = Path(f'{self.next_fresh():02}_model')
+                subdir = to_path.parent.joinpath(dirname)
+                topath = subdir.joinpath(Path(path).name)
+                mkdir(subdir)
+                self.model(base_path.joinpath(path), topath)
+                step['model'] = str(dirname.joinpath(Path(path).name))
+            elif 'data' in step:
+                path = step['data']
+                dirname = Path(f'{self.next_fresh():02}_data')
+                subdir = to_path.parent.joinpath(dirname)
+                topath = subdir.joinpath(Path(path).name)
+                mkdir(subdir)
+                self.data(base_path.joinpath(path), topath)
+                step['data'] = str(dirname.joinpath(Path(path).name))
+            elif 'nodegroups' in step:
+                path = step['nodegroups']
+                dirname = Path(f'{self.next_fresh():02}_nodegroups')
+                subdir = to_path.parent.joinpath(dirname)
+                mkdir(subdir)
+                self.nodegroups(base_path.joinpath(path), subdir)
+                step['nodegroups'] = str(dirname)
```

Comment:
Lots of duplicated if clauses above.  One possible alternative:

```
for attr in [ 'manifest', 'model', ''data', 'nodegroups' ]:
    if attr in step:
        path = step[attr]
        dirname = Path(f'{self.next_fresh():02}_{attr}')
        subdir = to_path.parent.joinpath(dirname)
        subdir.mkdir(parents=True, exist_ok=False)
        topath = subdir.joinpath(Path(path).name)
        # nodegroups is different for the following two; if that's intended, use if attr == "nodegroups" just here...
        getattr(self, attr)(base_path.joinpath(path), topath))
        step[attr] = str(dirname.joinpath(path).name)
        break   # [optional]
```

Reply:
I agree that there's a bit of duplication here. I'm hesitant to try and factor out more of the commonality because I think the future holds more diversity in ingestion steps (if there's new ones to come) than less, and I don't want to make the code more complicated just to try and eliminate some of the duplication.